core/client/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java

```diff
@@ -0,0 +1,443 @@
+  public void cancel() throws IOException {
+    if (mClosed) {
+      return;
+    }
+
+    mLock.lock();
+    try {
+      mThrowable = new IOException("PacketWriter is cancelled.");
+      mBufferEmptyOrFail.signal();
+      mBufferNotFullOrFail.signal();
+      mDoneOrFail.signal();
+
+      // TODO(peis): Better support cancel so that we do not need to close the channel.
+      ChannelFuture future = mChannel.close().sync();
+      if (future.cause() != null) {
+        throw new IOException(future.cause());
+      }
+    } catch (InterruptedException e) {
+      throw Throwables.propagate(e);
+    } finally {
+      mLock.unlock();
+      mClosed = true;
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    mChannel.flush();
+
+    mLock.lock();
+    try {
+      while (true) {
+        if (mPosToWrite == mPosToQueue) {
+          return;
+        }
+        if (mThrowable != null) {
+          throw new IOException(mThrowable);
+        }
+        if (!mBufferEmptyOrFail.await(WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+          throw new IOException(
+              String.format("Timeout to flush packets to %d @ %s.", mId, mAddress));
+        }
+      }
+    } catch (InterruptedException e) {
+      throw Throwables.propagate(e);
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+
+    sendEOF();
+    mLock.lock();
+    try {
+      while (true) {
+        if (mDone) {
+          return;
+        }
+        try {
+          if (mThrowable != null) {
+            mChannel.close().sync();
+            throw new IOException(mThrowable);
+          }
+          if (!mDoneOrFail.await(WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+            mChannel.close().sync();
+            throw new IOException(String.format(
+                "Timeout to close the NettyPacketWriter (block: %d, address: %s).", mId,
+                mAddress));
+          }
+        } catch (InterruptedException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    } finally {
+      mLock.unlock();
+      if (mChannel.isOpen()) {
+        Preconditions.checkState(mChannel.pipeline().last() instanceof PacketWriteHandler);
+        mChannel.pipeline().removeLast();
+      }
+      BlockStoreContext.releaseNettyChannel(mAddress, mChannel);
+      mClosed = true;
+    }
+  }
+
+  /**
+   * Add {@link PacketWriteHandler} to the channel pipeline.
+   */
+  private void addHandler() {
+    ChannelPipeline pipeline = mChannel.pipeline();
+    if (!(pipeline.last() instanceof RPCMessageDecoder)) {
+      throw new RuntimeException(String.format("Channel pipeline has unexpected handlers %s.",
+          pipeline.last().getClass().getCanonicalName()));
+    }
+    mChannel.pipeline().addLast(new PacketWriteHandler());
+  }
+
+  /**
+   * @return true if there are too many bytes in flight
+   */
+  private boolean tooManyPacketsInFlight() {
```

Comment:
They are the same. But I think "tooManyPackets" is more accurate as MAX_PACKETS_IN_FLIGHT is the one that controls the threshold. And the unit of MAX_PACKETS_IN_FLIGHT  is "packet". 

How about  keeping this name to be consistent with other classes ?